File: iambically.go
package iambically

import (
	"fmt"
	"io"
	"net/http"
	"strconv"
	"time"

	"appengine"
	"appengine/datastore"
	"appengine/user"
)

type Entry struct {
	Committed time.Time
	Content   []byte
}

func init() {
	http.HandleFunc("/fetch", fetch)
	http.HandleFunc("/commit", commit)
	http.HandleFunc("/auth", auth)
	http.HandleFunc("/login", login)
	http.HandleFunc("/logout", logout)
}

func fetch(w http.ResponseWriter, r *http.Request) {
	c := appengine.NewContext(r)
	if checkAccount(c, w, r) {
		return
	}
	start_str := r.FormValue("start")
	q := datastore.NewQuery("Entry").Order("Committed").Limit(10)
	if start, err := strconv.ParseInt(start_str, 10, 64); err != nil {
		http.Error(w, err.Error(), http.StatusInternalServerError)
		return
	} else {
		q = q.Filter("Committed >", time.Unix(0, start))
	}
	entries := make([]Entry, 0, 10)
	keys, err := q.GetAll(c, &entries)
	if err != nil {
		http.Error(w, err.Error(), http.StatusInternalServerError)
		return
	}
	w.Header().Set("Content-Type", "text/plain")
	w.WriteHeader(http.StatusOK)
	fmt.Fprintf(w, "%d\n", len(entries))
	for i, value := range entries {
		fmt.Fprintf(w, "%d %d %s\n",
			len(value.Content), value.Committed.UnixNano(), keys[i].Encode())
		w.Write(value.Content)
	}
}

func commit(w http.ResponseWriter, r *http.Request) {
	c := appengine.NewContext(r)
	if checkAccount(c, w, r) {
		return
	}
	replaces_str := r.FormValue("replaces")
	var key *datastore.Key
	var err error
	if replaces_str != "" {
		key, err = datastore.DecodeKey(replaces_str)
		if err != nil {
			http.Error(w, err.Error(), http.StatusInternalServerError)
			return
		}
	} else {
		key = datastore.NewIncompleteKey(c, "Entry", nil)
	}
	g := Entry{
		Committed: time.Now(),
		Content:   []byte(r.FormValue("content")),
	}
	key, err = datastore.Put(c, key, &g)
	if err != nil {
		http.Error(w, err.Error(), http.StatusInternalServerError)
		return
	}
	w.Header().Set("Content-Type", "text/plain")
	w.WriteHeader(http.StatusOK)
	fmt.Fprintf(w, "%d %s\n", g.Committed.UnixNano(), key.Encode())
}

func auth(w http.ResponseWriter, r *http.Request) {
	c := appengine.NewContext(r)
	if checkAccount(c, w, r) {
		return
	}
	w.Header().Set("Content-Type", "text/plain")
	w.WriteHeader(http.StatusOK)
	io.WriteString(w, "ok\n")
}

func login(w http.ResponseWriter, r *http.Request) {
	c := appengine.NewContext(r)
	if user.Current(c) == nil {
		url, err := user.LoginURL(c, r.URL.String())
		if err != nil {
			http.Error(w, err.Error(), http.StatusInternalServerError)
			return
		}
		w.Header().Set("Location", url)
		w.WriteHeader(http.StatusFound)
		return
	}
	if checkAccount(c, w, r) {
		return
	}
	w.Header().Set("Content-Type", "text/plain")
	w.WriteHeader(http.StatusOK)
	io.WriteString(w, "ok\n")
}

func logout(w http.ResponseWriter, r *http.Request) {
	c := appengine.NewContext(r)
	if user.Current(c) != nil {
		url, err := user.LogoutURL(c, r.URL.String())
		if err != nil {
			http.Error(w, err.Error(), http.StatusInternalServerError)
			return
		}
		w.Header().Set("Location", url)
		w.WriteHeader(http.StatusFound)
		return
	}
	w.Header().Set("Content-Type", "text/plain")
	w.WriteHeader(http.StatusOK)
	io.WriteString(w, "ok\n")
}

func checkAccount(
	c appengine.Context, w http.ResponseWriter, r *http.Request) bool {
	u := user.Current(c)
	if u == nil {
		http.Error(w, "login required", http.StatusUnauthorized)
		return true
	} else if user.IsAdmin(c) || u.Email == "bradnelson@google.com" {
		return false
	} else {
		http.Error(w, "access denied", http.StatusForbidden)
		return true
	}
}

File: iambically.py
import datetime
import webapp2

from google.appengine.api import users
from google.appengine.ext import ndb


class Entry(ndb.Model):
  Committed = ndb.DateTimeProperty()
  Content = ndb.BlobProperty()

  def unixnano(self):
    epoch = datetime.datetime.utcfromtimestamp(0)
    return int((self.Committed - epoch).total_seconds() * 1000000000.0)


class Auth(webapp2.RequestHandler):
  def get(self):
    if not users.is_current_user_admin():
      return
    self.response.headers['Content-Type'] = 'text/plain'
    self.response.out.write('ok\n')


class Login(webapp2.RequestHandler):
  def get(self):
    if users.get_current_user() is None:
      self.redirect(users.create_login_url('/login'))
    else:
      self.response.headers['Content-Type'] = 'text/plain'
      self.response.out.write('ok\n')


class Logout(webapp2.RequestHandler):
  def get(self):
    if users.get_current_user() is None:
      self.response.headers['Content-Type'] = 'text/plain'
      self.response.out.write('ok\n')
    else:
      self.redirect(users.create_login_url('/logout'))


class Fetch(webapp2.RequestHandler):
  def post(self):
    if not users.is_current_user_admin():
      return
    start = datetime.datetime.utcfromtimestamp(
        int(self.request.get('start', '0')) / 1000000000.0)
    query = Entry.query(Entry.Committed > start)
    entries = query.fetch(10)
    self.response.headers['Content-Type'] = 'text/plain'
    self.response.out.write('%d\n' % len(entries))
    for e in entries:
      self.response.out.write('%d %d %s\n' % (
          len(e.Content), e.unixnano(), e.key.urlsafe()))
      self.response.out.write(e.Content)


class Commit(webapp2.RequestHandler):
  def post(self):
    if not users.is_current_user_admin():
      return
    replaces = self.request.get('replaces', '')
    if replaces:
      e = ndb.Key(urlsafe=replaces).get()
    else:
      e = Entry()
    e.Committed = datetime.datetime.utcnow()
    e.Content = str(self.request.get('content'))
    e.put()
    self.response.headers['Content-Type'] = 'text/plain'
    self.response.out.write('%d %s\n' % (e.unixnano(), e.key.urlsafe()))


app = webapp2.WSGIApplication([
    ('/fetch', Fetch),
    ('/commit', Commit),
    ('/auth', Auth),
    ('/login', Login),
    ('/logout', Logout),
], debug=False)
